# Домашнє завдання до теми "Python Basics"
В цьому ДЗ важливою умовою виконання завдань є те, що ми не імпортуємо жодних бібліотек, а робимо все на чистому Python, використовуючи лише базовий функціонал. Це необхідно аби відпрацювати роботу з базовими типами даних. А використовувати бібліотеки ми ще навчимось далі :)

# Робота з рядками

1.Запишіть у змінну `name` ваше імʼя.

In [3]:
name ='Nataliia'
print (name)

Nataliia


2.Запишіть у змінну `surname` ваше прізвище.

In [4]:
surname = 'Hordiienko'
print (surname)

Hordiienko


3.Сконкатенуйте імʼя та прізвище в одну фразу через пробіл. Запишіть результат у нову змінну `name_surname` та виведіть результат:

In [5]:
print (name+ ' ' +surname)

Nataliia Hordiienko


In [86]:
name_surname = name , surname
print  (name,surname)

Nataliia Hordiienko


Очікувана відповідь (тільки з вашими даними): 'Hanna Pylieva' 

4.Зробіть усі літери змінної `name_surname` заглавними використовуючи вбудовану функцію Python для рядків:

In [7]:
name_surname.upper()

'NATALIIA HORDIIENKO'

Очікувана відповідь (тільки з вашими даними): 'HANNA PYLIEVA'

5.Виведіть третю літеру вашого `name_surname`. Тут уважно, треба саме третя літера. Згадайте, з якого значення починається індексування значень в Python.

In [8]:
name_surname [2]

't'

Очікувана відповідь (тільки з вашими даними): 'n'

6.Виведіть першу та останню літеру вашого прізвища.

In [11]:
len (name_surname)

19

In [10]:
name_surname [0:19]

'Nataliia Hordiienko'

In [21]:
name_surname [0] + name_surname [18]

'No'

In [80]:
name_surname [0] + ' ' + name_surname [18]

'N o'

In [90]:
name_surname [0] /n name_surname [18]

SyntaxError: invalid syntax (151016628.py, line 1)

#### Можете написати як розділити букви щоб відповідь була ('N', 'o'). Чому при використанні /n код не працює, як правильно написати код з використанням /n та /t. Дякую :)

Очікувана відповідь (тільки з вашими даними): ('P', 'a')

7.Виведіть довжину вашого прізвища.

In [147]:
len (surname)

10

Очікувана відповідь (тільки з вашими даними): 7

8.Виведіть кількість літер `a` у столиці Мадагаскару `Antananarivo`. Важливо врахувати всі літери, як великі, так і маленькі.

In [1]:
s = "Antananarivo"
s.count('a') + s.count('A')

4

Очікувана відповідь: 4

# Робота з числами та операторами

9.Створіть змінні a, b, c та d цілочисельного типу, задайде їм значення відповідно 1, 2, 3, 4, та обчисліть з допомогою Python наступний вираз:  $$ \frac{(a+b) * c}{d} $$

In [47]:
def add_numbers(a, b, c, d):
    x1 = a + b
    y1 = x1 * c
    x2 = y1 / d
    return x2

In [65]:
add_numbers(1,2,3,4)

2.25

Очікувана відповідь: 2.25

10.На змінних, створених в попердньому завданні, обчисліть наступний вираз з допомогою Python: $$ (a^b + c^d)^2 $$

In [73]:
def add_numbers(a, b, c, d):
    x3 = a ** b
    x4 = c ** d
    x5 = x3 + x4
    x6 = x5 ** 2
    return x6

In [74]:
add_numbers(1,2,3,4)

6724

Очікувана відповідь: 6724

# Робота зі колекціями

У нас є наступний список:

In [103]:
l = [
    78, 12, 14, 27, 47, 28, 51, 46, 94, 37, 42, 20, 52, 33, 11, 11, 51,
    20, 81, 54, 86, 51, 38, 42, 49, 99, 76, 17,  9, 23, 97, 29, 29, 63,
    40, 11, 31, 76, 15, 21, 29, 75, 35,  2, 10, 71, 17, 10, 13,  7, 44,
    12, 19, 61, 13, 25, 62, 46, 10, 44, 75, 57, 20, 57,  3, 44, 16, 28,
    79, 57, 62, 70, 95,  8, 47, 82, 80, 45, 36, 44, 79, 95, 73, 42, 29,
    83, 80, 66, 56,  8, 73,  5, 99, 65, 65, 98, 17, 36, 16, 60
]

Будемо з ним працювати в наступних завданнях.

11.Виведіть 5 перших елементів списку

In [111]:
l[0:5]

[78, 12, 14, 27, 47]

Очікувана відповідь: [78, 12, 14, 27, 47]

12.Виведіть максимальний і мінімальний елемент у списку

In [119]:
max (l) 

99

In [122]:
min (l)

2

Очікувана відповідь: 

```
Max element  99
Max element  2
```

13.Чи є в списку елементи, які повторюються? Перевірте це програмно і з допомогою інструкцій `if ... else` виведіть на екран відповідь "Так, є елементи, які повторюються", або "Ні, всі елементи унікальні".
Важливо це зробити без використання імпортів будь-яких бібліотек, на чистому Python так би мовити.

In [126]:
list = 1
if  
    print ("Так, є елементи, які повторюються")
else 
    print ("Ні, всі елементи унікальні")
      

SyntaxError: invalid syntax (1009683842.py, line 2)

14.Для кожного унікального елементу списку `l` знайдіть, скільки разів він повторюється та запишіть результат у змінну `n_repeats`. Виведіть `n_repeats` на екран.

Очікуваний результат (три крапки - для скорочення, у вас буде виведений повний набір унікальних елементів): 
```
{2: 1,
 3: 1,
 5: 1,
    ...
 97: 1,
 98: 1,
 99: 2}
```

15.Знайдіть три найменші унікальні елементи в списку `l`.

Очікувана відповідь: [2, 3, 5]


16.Використовуючи `n_repeats` скажіть, скільки разів повторюється кожен з трьох найменших елементів в списку.

Очікувана відповідь:  {2: 1, 3: 1, 5: 1}

17.Створіть новий список `l2`, який містить кожен 11-й елемент списку `l`. Виведіть `l2` на екран.  
Памʼятаємо, щоб наприклад, перший 11-и1 елемент списку має індекс 10.

Очікуваний результат: `[42, 51, 29, 2, 13, 44, 80, 66, 16]`

18.Створіть новий список `l2_squared`, де кожен елемент списку `l2` є піднесеним до степіня 2. 

Очікуваний результат: `[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]`

19.Знайдіть суму всіх елементів списку `l2` та виведіть її на екран.

20.Знайдіть середнє значення елементів списку `l2`, збережіть результат у змінну `l2_mean` та виведіть новостворену змінну та ії тип на екран.

21.Створіть новий список `l2_gt_mean`, за наступним правилом: "якщо цей елемент спиcку `l2` є більше за `l2_mean`, то запишемо 1, інакше 0. Виведіть новий список на екран.

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

# Ураа!! Ви завершили перше домашнє завдання!!!  
Аби його здати - залийте ноутбук у гітхаб репозиторій, який ви створили у перші дні навчання на курсі, та додайте посилання на файл, заповнишви форму https://forms.gle/EvxoiJiBTkuxFmhV7   
Так тримати! 🚀🚀🚀
